### PR TITLE
User is able to register NGS Measurements in the data manager application

### PR DESCRIPTION
**Pull request by Steffengreiner**
**What was changed**
Provide NGS Measurement registration and validation including
1.) Backend JPA Implementation
2.) Registration and Validation in Dialog
3.) Updated Grid columns and search functionality

**Open Questions**

1.) Saving NGS or Proteomics measurement currently swallows ALL exceptions during the registration process and returns a nonused result object. 
2.) Should the indexI7 and IndexI5 be collected and stored as a list within the ngsMeasurment similiar to the isotope labeling in the pxp measurement registration
3.) The measurementcode is only contains the samplecode of the first sample is this intended?

### CODE REVIEW

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java [hunk 1]
```diff
@@ -42,28 +46,74 @@ public class NGSMeasurement implements MeasurementMetadata {
...
+  @Column(name = "instrument", columnDefinition = "longtext CHECK (json_valid(`instrument`))")
   private OntologyTerm instrument;
 
   @Convert(converter = MeasurementCode.MeasurementCodeConverter.class)
   private MeasurementCode measurementCode;
-
-  @ElementCollection
-  @CollectionTable(name = "measurement_samples", joinColumns = @JoinColumn(name = "measurement_id"))
-  @Column(name = "measured_sample")
+  @Column(name = "registration")
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java:54]: Probably should be something like registrationTime

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java [hunk 2]
```diff
@@ -74,33 +124,26 @@ private NGSMeasurement(ProjectId projectId, Collection<SampleId> sampleIds,
...
-      MeasurementCode measurementCode, OntologyTerm instrument) {
+      MeasurementCode measurementCode, Organisation organisation, NGSMethodMetadata method,
+      String comment) throws IllegalArgumentException {
     if (sampleIds.isEmpty()) {
       throw new IllegalArgumentException(
           "No sample ids provided. At least one sample id must provided for a measurement.");
     }
-    requireNonNull(instrument);
     requireNonNull(measurementCode);
+    requireNonNull(method);
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java:138]: It would be great to have an error message here
```suggestion
    requireNonNull(method, "method must not be null");
```

**Review by KochTobi — DISMISSED** [review]
Code LGTM

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java (hunk 1 above)
Comment (Steffengreiner) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java:54]: Good point!

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/validation/MeasurementNGSValidator.java [hunk 3]
```diff
@@ -52,12 +48,13 @@ public static boolean isNGS(Collection<String> properties) {
...
     }
-    for (String ngsProperty : NGS_PROPERTIES) {
+    for (NGS_PROPERTY ngsProperty : NGS_PROPERTY.values()) {
       var propertyFound = properties.stream()
-          .filter(property -> Objects.equals(property.toLowerCase(), ngsProperty)).findAny();
+          .filter(property -> Objects.equals(property.toLowerCase(), ngsProperty.label()))
+          .findAny();
       if (propertyFound.isEmpty()) {
         return false;
       }
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/application/measurement/validation/MeasurementNGSValidator.java:60]: You are going through a list and checking whether this list contains all elements, correct?
Why not map the enum values to the required labels and call containsAll instead? This way java handles the optimisation of the search.

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/validation/MeasurementProteomicsValidator.java [hunk 4]
```diff
@@ -197,6 +197,13 @@ ValidationResult validateMandatoryDataProvided(
       } else {
         validation = validation.combine(ValidationResult.successful(1));
       }
+      if (measurementMetadata.fractionName().isBlank()) {
```
Comment (KochTobi) [inline project-management/src/main/java/life/qbic/projectmanagement/application/measurement/validation/MeasurementProteomicsValidator.java:200]: can this throw a NullPointer Exception?

File: user-interface/src/main/resources/templates/valid_ngs_measurement_registration_sheet.xlsx
Comment (KochTobi) [inline user-interface/src/main/resources/templates/valid_ngs_measurement_registration_sheet.xlsx:1]: How to review this?

**Review by KochTobi — COMMENTED** [review]
I still have to try it out. Some remarks on the code

**Comment by KochTobi** [comment]
some tests are failing

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java [hunk 5]
```diff
@@ -42,28 +46,74 @@ public class NGSMeasurement implements MeasurementMetadata {
...
-  @ElementCollection
-  @CollectionTable(name = "measurement_samples", joinColumns = @JoinColumn(name = "measurement_id"))
-  @Column(name = "measured_sample")
+  @Column(name = "registrationTime")
+  private Instant registration;
+  @Embedded
+  private Organisation organisation;
+  @ElementCollection(targetClass = SampleId.class, fetch = FetchType.EAGER)
+  @CollectionTable(name = "ngs_measurement_samples", joinColumns = @JoinColumn(name = "measurement_id"))
   private Collection<SampleId> measuredSamples;
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java:60]: Yes if we don't fetch the samples eagerly, the class will not contain the samples in question during processing 🤔

File: project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java [hunk 6]
```diff
@@ -42,28 +46,74 @@ public class NGSMeasurement implements MeasurementMetadata {
...
+      MeasurementCode measurementCode,
+      Organisation organisation, NGSMethodMetadata method, String comment, Instant registration) {
+    evaluateMandatoryMetadata(
+        method); // throws IllegalArgumentException if required properties are missing
     measuredSamples = new ArrayList<>();
     measuredSamples.addAll(sampleIds);
+    this.id = measurementId;
     this.projectId = requireNonNull(projectId, "projectId must not be null");
-    this.instrument = instrument;
+    this.organisation = organisation;
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply project-management/src/main/java/life/qbic/projectmanagement/domain/model/measurement/NGSMeasurement.java]: Good point i provided null checks for all required values 👍

File: user-interface/src/main/resources/templates/valid_ngs_measurement_registration_sheet.xlsx
Comment (Steffengreiner) [reply user-interface/src/main/resources/templates/valid_ngs_measurement_registration_sheet.xlsx:1]: This is the new template sheet. The easiest way to test it is to download it and fill it and try to register measurements.

File: project-management/src/main/java/life/qbic/projectmanagement/application/measurement/validation/MeasurementNGSValidator.java (hunk 3 above)
Comment (Steffengreiner) [reply project-management/src/main/java/life/qbic/projectmanagement/application/measurement/validation/MeasurementNGSValidator.java]: Good point, IntelliJ recommended to wrap the thing in a hashset to avoid performance issues which was news to me 😅

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/measurement/MeasurementLookupImplementation.java [hunk 7]
```diff
@@ -111,6 +109,7 @@ private Specification<ProteomicsMeasurement> generateProteomicsFilterSpecificati
             ontologyNameContains,
             ontologyLabelContains,
             facilityContains,
+            fractionContains,
```
Comment (wow-such-code) [inline project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/measurement/MeasurementLookupImplementation.java:112]: why was this changed in this PR and what does it mean?

**Review by wow-such-code — CHANGES_REQUESTED** [review]
looks good and everything seems to work. please have a look at the remarks, anyway, I think it might make sense to take the new error handling from the ptx update PR, or at least be careful that this will not lead to errors when merging.

other than that special care should be taken concerning singular/plural name "sample id(s)" in the upload sheets and the validation methods. my understanding was this was plural due to pooling - but if that was wrong please check it again.

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/measurement/MeasurementLookupImplementation.java (hunk 7 above)
Comment (Steffengreiner) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/experiment/measurement/MeasurementLookupImplementation.java:112]: This is a specification that is responsible for the search within the fraction name of a proteomics measurement. It was commented out unnecessarily since the user might want to search for this property as well 🤔

**Review by wow-such-code — APPROVED** [review]
